File: engine_java/src/main/java/com/alibabacloud/polar_race/engine/common/EngineRace.java
package com.alibabacloud.polar_race.engine.common;

import com.alibabacloud.polar_race.engine.common.exceptions.EngineException;
import com.alibabacloud.polar_race.engine.common.impl.DBImpl;

public class EngineRace extends AbstractEngine {

    DBImpl db;

    @Override
    public void open(String path) throws EngineException {
//		System.out.println("=======================db open=========================");
//		SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");//设置日期格式
//		System.out.println(df.format(new Date()));// new Date()为获取当前系统时间
        db = new DBImpl(path);
    }

    @Override
    public void write(byte[] key, byte[] value) throws EngineException {
        db.write(key, value);
    }

    @Override
    public byte[] read(byte[] key) throws EngineException {
        return db.read(key);
    }

    @Override
    public void range(byte[] lower, byte[] upper, AbstractVisitor visitor) throws EngineException {
    }


    @Override
    public void close() {
//	    System.out.println("=======================db close======================");
//		SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");//设置日期格式
//		System.out.println(df.format(new Date()));// new Date()为获取当前系统时间
        db.close();
        db = null;
        System.gc();
    }

}

File: engine_cpp/engine_race/engine_race.cc
// Copyright [2018] Alibaba Cloud All rights reserved
#include <sys/stat.h>
#include "engine_race.h"

namespace polar_race {

    RetCode Engine::Open(const std::string &name, Engine **eptr) {
        return EngineRace::Open(name, eptr);
    }

    Engine::~Engine() {
    }

/*
 * Complete the functions below to implement you own engine
 */

// 1. Open engine
    RetCode EngineRace::Open(const std::string &name, Engine **eptr) {
        struct stat st = {0};
        if(stat(name.data(), &st) == -1) {
            mkdir(name.data(), 0777);
        }

        fprintf(stderr, "Opening database!\n");

        auto *engine_race = new EngineRace(name);

        *eptr = engine_race;
        return kSucc;
    }

// 2. Close engine
    EngineRace::~EngineRace() {
        delete pEngine;
        fprintf(stderr, "closing database\n");
    }

// 3. Write a key-value pair into engine
    RetCode EngineRace::Write(const PolarString &key, const PolarString &value) {
        pEngine->put(key, value);
        return kSucc;
    }

// 4. Read value of a key
    RetCode EngineRace::Read(const PolarString &key, std::string *value) {
        return pEngine->read(key, value);
    }

/*
 * NOTICE: Implement 'Range' in quarter-final,
 *         you can skip it in preliminary.
 */
// 5. Applies the given Vistor::Visit function to the result
// of every key-value pair in the key range [first, last),
// in order
// lower=="" is treated as a key before all keys in the database.
// upper=="" is treated as a key after all keys in the database.
// Therefore the following call will traverse the entire database:
//   Range("", "", visitor)
    RetCode EngineRace::Range(const PolarString &lower, const PolarString &upper,
                              Visitor &visitor) {
        return pEngine->range(lower, upper, visitor);
    }

}  // namespace polar_race
